source/sn_coap_protocol.c

```diff
@@ -1390,15 +1427,22 @@ static bool sn_coap_protocol_linked_list_blockwise_payload_compare_block_number(
  * \brief Removes current stored blockwise paylod from Linked list
  *****************************************************************************/
 
-static void sn_coap_protocol_linked_list_blockwise_payload_remove_oldest(struct coap_s *handle)
+static void sn_coap_protocol_linked_list_blockwise_payload_remove_oldest(struct coap_s *handle, uint8_t *token_ptr, uint8_t token_len)
 {
-    coap_blockwise_payload_s *removed_payload_ptr;
-
     /* Remove oldest node in Linked list*/
-    removed_payload_ptr = ns_list_get_first(&handle->linked_list_blockwise_received_payloads);
-
-    if (removed_payload_ptr != NULL) {
-        sn_coap_protocol_linked_list_blockwise_payload_remove(handle, removed_payload_ptr);
+    if (token_ptr) {
+        ns_list_foreach(coap_blockwise_payload_s, removed_payload_ptr, &handle->linked_list_blockwise_received_payloads) {
+            if ((token_len == removed_payload_ptr->token_len) && !memcmp(removed_payload_ptr->token_ptr, token_ptr, token_len)) {
+                sn_coap_protocol_linked_list_blockwise_payload_remove(handle, removed_payload_ptr);
+                return;
+            }
+        }
+    } else {
+        ns_list_foreach(coap_blockwise_payload_s, removed_payload_ptr, &handle->linked_list_blockwise_received_payloads) {
```

Comment:
This will now cleanup the whole list, why not to use ns_list_get_first() still?

Reply:
Now it is removing first payload without token. Added return to loop when removed_payload_ptr is found.